pkg/ruler/api.go

```diff
@@ -0,0 +1,264 @@
+	Rules    []rule  `json:"rules"`
+	Interval float64 `json:"interval"`
+}
+
+type rule interface{}
+
+type alertingRule struct {
+	// State can be "pending", "firing", "inactive".
+	State       string        `json:"state"`
+	Name        string        `json:"name"`
+	Query       string        `json:"query"`
+	Duration    float64       `json:"duration"`
+	Labels      labels.Labels `json:"labels"`
+	Annotations labels.Labels `json:"annotations"`
+	Alerts      []*Alert      `json:"alerts"`
+	Health      string        `json:"health"`
+	LastError   string        `json:"lastError,omitempty"`
+	// Type of an alertingRule is always "alerting".
+	Type string `json:"type"`
+}
+
+type recordingRule struct {
+	Name      string        `json:"name"`
+	Query     string        `json:"query"`
+	Labels    labels.Labels `json:"labels,omitempty"`
+	Health    string        `json:"health"`
+	LastError string        `json:"lastError,omitempty"`
+	// Type of a recordingRule is always "recording".
+	Type string `json:"type"`
+}
+
+func respondError(logger log.Logger, w http.ResponseWriter, msg string, errType errorType) {
+	b, err := json.Marshal(&response{
+		Status:    "error",
+		ErrorType: errType,
+		Error:     msg,
+		Data:      nil,
+	})
+
+	if err != nil {
+		level.Error(logger).Log("msg", "error marshaling json response", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusInternalServerError)
+	if n, err := w.Write(b); err != nil {
+		level.Error(logger).Log("msg", "error writing response", "bytesWritten", n, "err", err)
+	}
+}
+
+func (r *Ruler) rules(w http.ResponseWriter, req *http.Request) {
+	logger := util.WithContext(req.Context(), util.Logger)
+	userID, ctx, err := user.ExtractOrgIDFromHTTPRequest(req)
+	if err != nil {
+		level.Error(logger).Log("msg", "error extracting org id from context", "err", err)
+		respondError(logger, w, "no valid org id found", errorInternal)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	rgs, err := r.GetRules(ctx, userID)
+
+	if err != nil {
+		respondError(logger, w, err.Error(), errorInternal)
+		return
+	}
+
+	groups := make([]*RuleGroup, len(rgs))
+
+	for _, g := range rgs {
+		grp := RuleGroup{
+			Name:     g.Name,
+			File:     g.Namespace,
+			Interval: g.Interval.Seconds(),
+			Rules:    make([]rule, len(g.Rules)),
+		}
+
+		for i, rl := range g.Rules {
+			if g.Rules[i].Alert != "" {
+				alerts := make([]*Alert, len(rl.Alerts))
+				for _, a := range rl.Alerts {
+					alerts = append(alerts, &Alert{
+						Labels:      client.FromLabelAdaptersToLabels(a.Labels),
+						Annotations: client.FromLabelAdaptersToLabels(a.Annotations),
+						State:       a.GetState(),
+						ActiveAt:    a.ActiveAt,
+						Value:       fmt.Sprintf("%f", a.Value),
+					})
+				}
+				grp.Rules[i] = alertingRule{
+					State:       rl.GetState(),
+					Name:        rl.GetAlert(),
+					Query:       rl.GetExpr(),
+					Duration:    rl.For.Seconds(),
```

Comment:
When the duration is not set in the alert, this is exported as `0`. Is it correct? May you check Prometheus behaviour please?

Reply:
I double checked and Prometheus exposes it as `0`, so all good here.

Reply:
Yes, in the case when `For` is not set the alert will fire immediately. From the [Prometheus Docs](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/#defining-alerting-rules):

> The optional for clause causes Prometheus to wait for a certain duration between first encountering a new expression output vector element and counting an alert as firing for this element. In this case, Prometheus will check that the alert continues to be active during each evaluation for 10 minutes before firing the alert. Elements that are active, but not firing yet, are in the pending state.